Support Panache Sort Null Handling

Sgitario opened the pull request:
Allow users sorting by either nulls first or nulls last.
Example of usage:

```java
Sort.by("foo", Sort.Direction.Ascending, Sort.NullPrecedence.NULLS_FIRST);
```

(See more examples in tests)

This PR also adds support of nulls handling for the Spring Data JPA Quarkus extension: it translates the Spring Sort object to the new Panache Sort object. 

Finally, as the Panache Sort API is also used by the Mongo Quarkus extension and Mongo does not easily support sorting by nulls, I decided to simply print a WARNING message if users try to use it.

Fix https://github.com/quarkusio/quarkus/issues/26172

loicmathieu commented:
I like the idea, I'll try to review the PR tomorrow or Friday.

On Mongo, we should not create methods that allow to set the NullHandling if not supported, or we should throw an error not log a warning as it's a developer mistake.

Sgitario commented:
> I like the idea, I'll try to review the PR tomorrow or Friday.
> 
> On Mongo, we should not create methods that allow to set the NullHandling if not supported, or we should throw an error not log a warning as it's a developer mistake.

The methods are common for the non-Mongo extensions, but it makes sense to throw an exception instead.

gastaldi commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -40,9 +40,24 @@ public enum Direction {
+    /**
+     * Represents the order of null columns.
+     */
+    public enum NullHandling {
```
Maybe it would be nice to use the same name as the Hibernate Dialect uses: `NullPrecedence`

Sgitario replied:
Good idea

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -100,6 +129,20 @@ public static Sort by(String column, Direction direction) {
+     * @see #by(String)
+     * @see #by(String...)
+     */
+    public static Sort by(String column, Direction direction, NullHandling nullHandling) {
```
Maybe also add `by(String, NullHandling)`

Sgitario replied:
I'm a bit against about this change as internally, we would need to use this constructor:
```java
public Column(String name, Direction direction, NullPrecedence nullPrecedence) {
```
So, we need a direction. We could use the default direction ASCENDING tho.
wdyt?

loicmathieu replied:
Yes, use the default.

Sgitario replied:
Done

gastaldi reviewed (approved):
LGTM

gastaldi commented:
I think it would be nice to have this new feature documented in the Panache guides too.

gastaldi commented:
Also don't forget to change the snippet in the commit to use `NullPrecendence` instead of `NullHandling` 😉

loicmathieu reviewed (changes requested):
Can you also add some tests for MongoDB ? In existing test cases.

Sgitario commented:
| Can you also add some tests for MongoDB ? In existing test cases.

@loicmathieu do you mean to verify that the exception is thrown when using NullPrecedence in MongoDB ?

loicmathieu commented:
@Sgitario yes

Sgitario commented on extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/PanacheSortTestCase.java:
```diff
@@ -0,0 +1,30 @@
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class PanacheSortTestCase {
```
Moved to `PanacheHotReloadTestCase`

Sgitario commented:
I've just updated the PR's changes with:
- documentation
- integration tests in reactive, orm and mongodb

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -100,6 +129,33 @@ public static Sort by(String column, Direction direction) {
+    }
+
+    /**
+     * Sort by the given column, in the given order and with null handling.
```
```suggestion
     * Sort by the given column, in the given order and  in the given null precedence.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -100,6 +129,33 @@ public static Sort by(String column, Direction direction) {
+     *
+     * @param column the column to sort on, in the given order.
+     * @param direction the direction to sort on.
+     * @param nullPrecedence the null handling to use.
```
```suggestion
     * @param nullPrecedence the null precedence to use.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -100,6 +129,33 @@ public static Sort by(String column, Direction direction) {
+     * @param column the column to sort on, in the given order.
+     * @param direction the direction to sort on.
+     * @param nullPrecedence the null handling to use.
+     * @return a new Sort instance which sorts on the given column in the given order.
```
```suggestion
     * @return a new Sort instance which sorts on the given column in the given order and null precedence.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -210,6 +267,32 @@ public Sort and(String name, Direction direction) {
     }
 
+    /**
+     * Adds a sort column, in the given order.
```
```suggestion
     * Adds a sort column, in the given null precedence.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -210,6 +267,32 @@ public Sort and(String name, Direction direction) {
+    /**
+     * Adds a sort column, in the given order.
+     *
+     * @param name the new column to sort on, in the given order.
```
```suggestion
     * @param name the new column to sort on, in the given null precedence.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -210,6 +267,32 @@ public Sort and(String name, Direction direction) {
+     * Adds a sort column, in the given order.
+     *
+     * @param name the new column to sort on, in the given order.
+     * @param nullPrecedence the null handling to use.
```
```suggestion
     * @param nullPrecedence the null precedence to use.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -210,6 +267,32 @@ public Sort and(String name, Direction direction) {
+    }
+
+    /**
+     * Adds a sort column, in the given order.
```
```suggestion
     * Adds a sort column, in the given order and null precedence.
```

loicmathieu commented on extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/Sort.java:
```diff
@@ -210,6 +267,32 @@ public Sort and(String name, Direction direction) {
+    /**
+     * Adds a sort column, in the given order.
+     *
+     * @param name the new column to sort on, in the given order.
```
```suggestion
     * @param name the new column to sort on, in the given order and null precedence.
```

loicmathieu reviewed (changes requested):
I added a bunch of JavaDoc suggestions, mainly due to the renaming of handlin -> precedence but overall looks good to me.

Can you please also update the Hibernate reactive guide Sort section: https://quarkus.io/guides/hibernate-reactive-panache#sorting

Sgitario commented:
Documentation for Panache Reactive also updated.

Sgitario committed:
Support Panache Sort Null Precedence

Allow users sorting by either nulls first or nulls last.
Example of usage:

```java
Sort.by("foo", Sort.Direction.Ascending, Sort.NullPrecedence.NULLS_FIRST);
```

(See more examples in tests)

This PR also adds support of nulls handling for the Spring Data JPA Quarkus extension: it translates the Spring Sort object to the new Panache Sort object. 

Finally, as the Panache Sort API is also used by the Mongo Quarkus extension and Mongo does not easily support sorting by nulls, I decided to simply print a WARNING message if users try to use it.

Fix https://github.com/quarkusio/quarkus/issues/26172

loicmathieu reviewed (approved):
LGTM, thanks for your contribution
